File: tools/fleet-registry/fleet_registry.py
"""Read-only loader for registry/fleet.yaml — the canonical fleet site list.

Consumers should ask this module who exists instead of keeping their own
roster. Tool-specific settings stay in the tool's own config, keyed by domain,
as an *overlay* on this list.

    from fleet_registry import load, sites, get, with_capability

    for domain in sites(status="live"):
        ...
    for domain in with_capability("analytics"):
        ...
    entry = get("totaljerks.com")   # dict, or None

The registry is cached after the first read; pass ``fresh=True`` to reload.
"""

from __future__ import annotations

import os
from pathlib import Path

import yaml

_DEFAULT_ROOT = Path(__file__).resolve().parents[2]
_cache: dict[str, dict] = {}


def registry_path(root: str | os.PathLike | None = None) -> Path:
    """Resolve the registry path. Container roots mount the repo at /work."""
    if root:
        return Path(root) / "registry" / "fleet.yaml"
    env = os.environ.get("FLEET_REGISTRY")
    if env:
        return Path(env)
    for candidate in (_DEFAULT_ROOT, Path("/work")):
        path = candidate / "registry" / "fleet.yaml"
        if path.exists():
            return path
    return _DEFAULT_ROOT / "registry" / "fleet.yaml"


def load(root: str | os.PathLike | None = None, fresh: bool = False) -> dict[str, dict]:
    """Return ``{domain: entry}`` for every registered site."""
    path = registry_path(root)
    key = str(path)
    if fresh or key not in _cache:
        data = yaml.safe_load(path.read_text(encoding="utf-8")) or {}
        _cache[key] = data.get("sites") or {}
    return _cache[key]


def sites(status: str | None = None, root=None) -> list[str]:
    """Domains, optionally filtered by status (live / scaffold / parked / redirect)."""
    entries = load(root)
    return sorted(d for d, e in entries.items() if status is None or e.get("status") == status)


def get(domain: str, root=None) -> dict | None:
    return load(root).get(domain)


def with_capability(capability: str, root=None) -> list[str]:
    """Domains a given fleet system applies to (site, ops, cron, smoke, tasks,
    affiliate, analytics, social, data-hub, product-feed)."""
    entries = load(root)
    return sorted(
        d for d, e in entries.items()
        if capability in (e.get("capabilities_override") or e.get("capabilities") or [])
    )


def repo(domain: str, root=None) -> str | None:
    entry = get(domain, root) or {}
    return entry.get("repo")


def analytics(domain: str, root=None) -> dict:
    entry = get(domain, root) or {}
    return entry.get("analytics") or {}

File: tools/fleet-registry/fleet-registry.js
'use strict';

// Read-only loader for registry/fleet.yaml — the canonical fleet site list.
//
// Node twin of fleet_registry.py. Consumers (fleet-dashboard, post-notify,
// content-guardrails) ask this who exists instead of keeping their own roster;
// tool-specific settings stay in the tool's config keyed by domain.
//
//   const reg = require('../fleet-registry/fleet-registry');
//   reg.sites(root, { status: 'live' })
//   reg.withCapability(root, 'cron')
//   reg.get(root, 'totaljerks.com')
//
// Mirrors sites.js's short-TTL memoization so hot routes don't re-parse YAML.

const fs = require('node:fs');
const path = require('node:path');
const yaml = require('js-yaml');

const TTL_MS = 5000;
const _cache = new Map(); // path -> { at, sites }

function registryPath(root) {
  if (process.env.FLEET_REGISTRY) return process.env.FLEET_REGISTRY;
  return path.join(root || path.resolve(__dirname, '..', '..'), 'registry', 'fleet.yaml');
}

function load(root, { fresh = false } = {}) {
  const file = registryPath(root);
  const hit = _cache.get(file);
  if (!fresh && hit && Date.now() - hit.at < TTL_MS) return hit.sites;
  let sites = {};
  try {
    sites = (yaml.load(fs.readFileSync(file, 'utf8')) || {}).sites || {};
  } catch {
    sites = {};
  }
  _cache.set(file, { at: Date.now(), sites });
  return sites;
}

function sites(root, { status = null, fresh = false } = {}) {
  const entries = load(root, { fresh });
  return Object.keys(entries)
    .filter(d => status === null || entries[d].status === status)
    .sort();
}

function get(root, domain) {
  return load(root)[domain] || null;
}

function withCapability(root, capability) {
  const entries = load(root);
  return Object.keys(entries)
    .filter(d =>
      (entries[d].capabilities_override || entries[d].capabilities || []).includes(capability)
    )
    .sort();
}

module.exports = { load, sites, get, withCapability, registryPath };
